File: lib/bot/rustrelease.go
package bot

import (
	"fmt"
	"time"

	tgb "gopkg.in/telebot.v3"
)

const (
	epochRelease = 5
	dtFmt        = "Jan 02 2006"
)

var epoch = time.Date(2015, time.December, 10, 0, 0, 0, 0, time.UTC)

type rustV1Release struct{ date time.Time }

func (r rustV1Release) minor() int {
	weeksSinceEpoch := int(r.date.Sub(epoch).Hours()) / (24 * 7)
	if weeksSinceEpoch < 0 {
		return -1
	}
	newReleases := weeksSinceEpoch / 6
	return epochRelease + newReleases
}

func (r rustV1Release) releaseDate() time.Time {
	newReleases := r.minor() - epochRelease
	return epoch.AddDate(0, 0, newReleases*6*7)
}

func (r rustV1Release) String() string {
	return fmt.Sprintf("Rust v1.%d\t(%s)", r.minor(), r.releaseDate().Format(dtFmt))
}

func RustRelease(ctx tgb.Context) error {
	// Based on https://forge.rust-lang.org/js/index.js.
	now := time.Now()

	stable := rustV1Release{now}
	beta := rustV1Release{now.AddDate(0, 0, 7*6)}
	nightly := rustV1Release{now.AddDate(0, 0, 7*6*2)}
	next := rustV1Release{now.AddDate(0, 0, 7*6*3)}

	return ctx.Reply(fmt.Sprintf(
		"Oh, I just asked Ferris 🦀️...\n```\nstable:\t%v\nbeta:\t%v\nnightly:\t%v\nnext:\t%v```\n",
		stable, beta, nightly, next,
	), &tgb.SendOptions{ParseMode: tgb.ModeMarkdown})
}

File: lib/angela/command/rust_release.ex
defmodule Angela.Command.RustRelease do
  @moduledoc """
  Shows current and upcoming Rust release versions based on the 6-week release cycle.
  """

  alias Angela.Command.Response
  alias ExGram.Model.{Message, ReplyParameters}

  @epoch ~D[2015-12-10]
  @epoch_release 5
  @date_fmt "%b %d %Y"

  @behaviour Angela.Command

  @impl true
  def usage(), do: "/rustrelease"

  @impl true
  def respond(msg = %Message{}) do
    # Convert Unix timestamp to Date
    now = msg.date |> DateTime.from_unix!() |> DateTime.to_date()

    # Based on https://forge.rust-lang.org/js/index.js
    stable = rust_v1_release(now)
    beta = rust_v1_release(Date.add(now, 7 * 6))
    nightly = rust_v1_release(Date.add(now, 7 * 6 * 2))
    next = rust_v1_release(Date.add(now, 7 * 6 * 3))

    """
    Oh, I just asked Ferris 🦀️...
    ```
    stable:\t#{stable}
    beta:\t#{beta}
    nightly:\t#{nightly}
    next:\t#{next}
    ```\
    """
    |> Response.new(
      reply_parameters: %ReplyParameters{message_id: msg.message_id},
      parse_mode: "Markdown"
    )
  end

  defp rust_v1_release(date) do
    date_str =
      (minor = date |> minor_version())
      |> release_date()
      |> Calendar.strftime(@date_fmt)

    "Rust v1.#{minor}\t(#{date_str})"
  end

  defp minor_version(date) do
    days = Date.diff(date, @epoch)

    if days >= 0 do
      weeks = div(days, 7)
      @epoch_release + div(weeks, 6)
    else
      -1
    end
  end

  defp release_date(minor_version) do
    new_releases = minor_version - @epoch_release
    Date.add(@epoch, new_releases * 6 * 7)
  end
end
